In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/stanford-earthquake-dataset-stead/merge.hdf5
/kaggle/input/stanford-earthquake-dataset-stead/merge.csv


Para trabajar con el dataset, utilizamos los servicios de kaggle que nos permiten guardar el input de manera persistente, lo cual es conveniente al ser un dataset de ~90GB.

In [2]:
import os
import pandas as pd

input_dir = '/kaggle/input/stanford-earthquake-dataset-stead/'
filename = 'merge.csv'
df_original = pd.read_csv(os.path.join(input_dir, filename), low_memory=False)
df_original.head()

network_code receiver_code receiver_type  receiver_latitude  \
0           TA          109C            HH            32.8889   
1           TA          109C            HH            32.8889   
2           TA          109C            HH            32.8889   
3           TA          109C            HH            32.8889   
4           TA          109C            HH            32.8889   

   receiver_longitude  receiver_elevation_m  p_arrival_sample p_status  \
0           -117.1051                 150.0               NaN      NaN   
1           -117.1051                 150.0               NaN      NaN   
2           -117.1051                 150.0               NaN      NaN   
3           -117.1051                 150.0               NaN      NaN   
4           -117.1051                 150.0               NaN      NaN   

   p_weight  p_travel_sec  ...  source_magnitude_author  \
0       NaN           NaN  ...                      NaN   
1       NaN           NaN  ...                      NaN   
2       NaN           NaN  ...                      NaN   
3       NaN           NaN  ...                      NaN   
4       NaN           NaN  ...                      NaN   

  source_mechanism_strike_dip_rake  source_distance_deg source_distance_km  \
0                              NaN                  NaN                NaN   
1                              NaN                  NaN                NaN   
2                              NaN                  NaN                NaN   
3                              NaN                  NaN                NaN   
4                              NaN                  NaN                NaN   

  back_azimuth_deg  snr_db  coda_end_sample     trace_start_time  \
0              NaN     NaN              NaN  2015-10-21 05:55:00   
1              NaN     NaN              NaN  2015-11-06 14:50:00   
2              NaN     NaN              NaN  2015-11-07 02:20:00   
3              NaN     NaN              NaN  2015-11-14 05:15:00   
4              NaN     NaN              NaN  2015-12-25 18:50:00   

   trace_category               trace_name  
0           noise  109C.TA_201510210555_NO  
1           noise  109C.TA_201511061450_NO  
2           noise  109C.TA_201511070220_NO  
3           noise  109C.TA_201511140515_NO  
4           noise  109C.TA_201512251850_NO  

[5 rows x 35 columns]

Realizamos una copia del dataset, para ejecutar desde este punto las celdas, y no tener que volver a leer el dataset a pandas

In [3]:
df = df_original.copy()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1265657 entries, 0 to 1265656
Data columns (total 35 columns):
 #   Column                            Non-Null Count    Dtype  
---  ------                            --------------    -----  
 0   network_code                      1265613 non-null  object 
 1   receiver_code                     1265657 non-null  object 
 2   receiver_type                     1265657 non-null  object 
 3   receiver_latitude                 1265657 non-null  float64
 4   receiver_longitude                1265657 non-null  float64
 5   receiver_elevation_m              1265657 non-null  float64
 6   p_arrival_sample                  1030231 non-null  float64
 7   p_status                          1030231 non-null  object 
 8   p_weight                          1030057 non-null  float64
 9   p_travel_sec                      1030231 non-null  float64
 10  s_arrival_sample                  1030231 non-null  float64
 11  s_status                          103

Se trata de un dataset con 1265657 filas y 35 columnas.

La variable trace_name es un identificador único para cada observación:

In [4]:
len(df['trace_name'].unique()) == len(df)

True

Por tanto, lo podemos utilizar como índice de fila.

In [5]:
df = df.set_index('trace_name')
df.head()

network_code receiver_code receiver_type  \
trace_name                                                         
109C.TA_201510210555_NO           TA          109C            HH   
109C.TA_201511061450_NO           TA          109C            HH   
109C.TA_201511070220_NO           TA          109C            HH   
109C.TA_201511140515_NO           TA          109C            HH   
109C.TA_201512251850_NO           TA          109C            HH   

                         receiver_latitude  receiver_longitude  \
trace_name                                                       
109C.TA_201510210555_NO            32.8889           -117.1051   
109C.TA_201511061450_NO            32.8889           -117.1051   
109C.TA_201511070220_NO            32.8889           -117.1051   
109C.TA_201511140515_NO            32.8889           -117.1051   
109C.TA_201512251850_NO            32.8889           -117.1051   

                         receiver_elevation_m  p_arrival_sample p_status  \
trace_name                                                                 
109C.TA_201510210555_NO                 150.0               NaN      NaN   
109C.TA_201511061450_NO                 150.0               NaN      NaN   
109C.TA_201511070220_NO                 150.0               NaN      NaN   
109C.TA_201511140515_NO                 150.0               NaN      NaN   
109C.TA_201512251850_NO                 150.0               NaN      NaN   

                         p_weight  p_travel_sec  ...  source_magnitude_type  \
trace_name                                       ...                          
109C.TA_201510210555_NO       NaN           NaN  ...                    NaN   
109C.TA_201511061450_NO       NaN           NaN  ...                    NaN   
109C.TA_201511070220_NO       NaN           NaN  ...                    NaN   
109C.TA_201511140515_NO       NaN           NaN  ...                    NaN   
109C.TA_201512251850_NO       NaN           NaN  ...                    NaN   

                        source_magnitude_author  \
trace_name                                        
109C.TA_201510210555_NO                     NaN   
109C.TA_201511061450_NO                     NaN   
109C.TA_201511070220_NO                     NaN   
109C.TA_201511140515_NO                     NaN   
109C.TA_201512251850_NO                     NaN   

                         source_mechanism_strike_dip_rake source_distance_deg  \
trace_name                                                                      
109C.TA_201510210555_NO                               NaN                 NaN   
109C.TA_201511061450_NO                               NaN                 NaN   
109C.TA_201511070220_NO                               NaN                 NaN   
109C.TA_201511140515_NO                               NaN                 NaN   
109C.TA_201512251850_NO                               NaN                 NaN   

                        source_distance_km  back_azimuth_deg  snr_db  \
trace_name                                                             
109C.TA_201510210555_NO                NaN               NaN     NaN   
109C.TA_201511061450_NO                NaN               NaN     NaN   
109C.TA_201511070220_NO                NaN               NaN     NaN   
109C.TA_201511140515_NO                NaN               NaN     NaN   
109C.TA_201512251850_NO                NaN               NaN     NaN   

                         coda_end_sample     trace_start_time  trace_category  
trace_name                                                                     
109C.TA_201510210555_NO              NaN  2015-10-21 05:55:00           noise  
109C.TA_201511061450_NO              NaN  2015-11-06 14:50:00           noise  
109C.TA_201511070220_NO              NaN  2015-11-07 02:20:00           noise  
109C.TA_201511140515_NO              NaN  2015-11-14 05:15:00           noise  
109C.TA_201512251850_NO              NaN  2015-12-25 18:50:00           noise  

[5 ro

network_code, receiver_code, trace_start_time y trace_category se especifican en el trace_name  

trace_name = [receiver_code].[network_code]\_[trace_start_time]\_[trace_category]

Observaciones:
- trace_start_time se expresa como unión sin espacios de las unidades de tiempo
- trace_category toma valores NO y EV para categorías noise y earthquake_local respectivamente

Se convierten a formato fecha las variables source_origin_time y trace_start_time

In [6]:
df['trace_start_time'] = pd.to_datetime(df['trace_start_time'], format='ISO8601')# format="%Y-%m-%d %H:%M:%S")
df['source_origin_time'] = pd.to_datetime(df['source_origin_time'], format='ISO8601')

In [7]:
df.iloc[0]['trace_start_time']

Timestamp('2015-10-21 05:55:00')

Se han recogido dos tipos de señales:  
- noise: sonidos no debidos a terremotos
- earthquake_local: terremotos que se encuentran en un radio menor a 350km

In [8]:
df.loc[:, 'trace_category'].unique()

array(['noise', 'earthquake_local'], dtype=object)

Para estudios que vamos a realizar, vamos dividir en dataframes uno por cada tipo de señal:

### Señales noise

Obtenemos un subdataset cuyas observaciones son señales noise, y podemos ver el ratio y el tamaño del subconjunto

In [9]:
df_noise = df.loc[df.loc[:, 'trace_category'] == 'noise']
size = len(df_noise)
print(f'len(df_noise)={size}, ratio={size/len(df)*100:.2f}%')
df_noise.head()

len(df_noise)=235426, ratio=18.60%


network_code receiver_code receiver_type  \
trace_name                                                         
109C.TA_201510210555_NO           TA          109C            HH   
109C.TA_201511061450_NO           TA          109C            HH   
109C.TA_201511070220_NO           TA          109C            HH   
109C.TA_201511140515_NO           TA          109C            HH   
109C.TA_201512251850_NO           TA          109C            HH   

                         receiver_latitude  receiver_longitude  \
trace_name                                                       
109C.TA_201510210555_NO            32.8889           -117.1051   
109C.TA_201511061450_NO            32.8889           -117.1051   
109C.TA_201511070220_NO            32.8889           -117.1051   
109C.TA_201511140515_NO            32.8889           -117.1051   
109C.TA_201512251850_NO            32.8889           -117.1051   

                         receiver_elevation_m  p_arrival_sample p_status  \
trace_name                                                                 
109C.TA_201510210555_NO                 150.0               NaN      NaN   
109C.TA_201511061450_NO                 150.0               NaN      NaN   
109C.TA_201511070220_NO                 150.0               NaN      NaN   
109C.TA_201511140515_NO                 150.0               NaN      NaN   
109C.TA_201512251850_NO                 150.0               NaN      NaN   

                         p_weight  p_travel_sec  ...  source_magnitude_type  \
trace_name                                       ...                          
109C.TA_201510210555_NO       NaN           NaN  ...                    NaN   
109C.TA_201511061450_NO       NaN           NaN  ...                    NaN   
109C.TA_201511070220_NO       NaN           NaN  ...                    NaN   
109C.TA_201511140515_NO       NaN           NaN  ...                    NaN   
109C.TA_201512251850_NO       NaN           NaN  ...                    NaN   

                        source_magnitude_author  \
trace_name                                        
109C.TA_201510210555_NO                     NaN   
109C.TA_201511061450_NO                     NaN   
109C.TA_201511070220_NO                     NaN   
109C.TA_201511140515_NO                     NaN   
109C.TA_201512251850_NO                     NaN   

                         source_mechanism_strike_dip_rake source_distance_deg  \
trace_name                                                                      
109C.TA_201510210555_NO                               NaN                 NaN   
109C.TA_201511061450_NO                               NaN                 NaN   
109C.TA_201511070220_NO                               NaN                 NaN   
109C.TA_201511140515_NO                               NaN                 NaN   
109C.TA_201512251850_NO                               NaN                 NaN   

                        source_distance_km  back_azimuth_deg  snr_db  \
trace_name                                                             
109C.TA_201510210555_NO                NaN               NaN     NaN   
109C.TA_201511061450_NO                NaN               NaN     NaN   
109C.TA_201511070220_NO                NaN               NaN     NaN   
109C.TA_201511140515_NO                NaN               NaN     NaN   
109C.TA_201512251850_NO                NaN               NaN     NaN   

                         coda_end_sample    trace_start_time  trace_category  
trace_name                                                                    
109C.TA_201510210555_NO              NaN 2015-10-21 05:55:00           noise  
109C.TA_201511061450_NO              NaN 2015-11-06 14:50:00           noise  
109C.TA_201511070220_NO              NaN 2015-11-07 02:20:00           noise  
109C.TA_201511140515_NO              NaN 2015-11-14 05:15:00           noise  
109C.TA_201512251850_NO              NaN 2015-12-25 18:50:00           noise  

[5 rows x 34

En la documentación del dataset se muestra que las señales categorizadas como noise, solo cuentan con 8 variables y el resto toman valores NA, eliminamos las variables que tienen todos sus valores NA:

In [10]:
df_noise = df_noise.dropna(axis=1, how='all')
print(df_noise.shape)
df_noise.head()

(235426, 8)


network_code receiver_code receiver_type  \
trace_name                                                         
109C.TA_201510210555_NO           TA          109C            HH   
109C.TA_201511061450_NO           TA          109C            HH   
109C.TA_201511070220_NO           TA          109C            HH   
109C.TA_201511140515_NO           TA          109C            HH   
109C.TA_201512251850_NO           TA          109C            HH   

                         receiver_latitude  receiver_longitude  \
trace_name                                                       
109C.TA_201510210555_NO            32.8889           -117.1051   
109C.TA_201511061450_NO            32.8889           -117.1051   
109C.TA_201511070220_NO            32.8889           -117.1051   
109C.TA_201511140515_NO            32.8889           -117.1051   
109C.TA_201512251850_NO            32.8889           -117.1051   

                         receiver_elevation_m    trace_start_time  \
trace_name                                                          
109C.TA_201510210555_NO                 150.0 2015-10-21 05:55:00   
109C.TA_201511061450_NO                 150.0 2015-11-06 14:50:00   
109C.TA_201511070220_NO                 150.0 2015-11-07 02:20:00   
109C.TA_201511140515_NO                 150.0 2015-11-14 05:15:00   
109C.TA_201512251850_NO                 150.0 2015-12-25 18:50:00   

                        trace_category  
trace_name                              
109C.TA_201510210555_NO          noise  
109C.TA_201511061450_NO          noise  
109C.TA_201511070220_NO          noise  
109C.TA_201511140515_NO          noise  
109C.TA_201512251850_NO          noise

Una vez hecho esto, podemos comprobar si existen más valores NA, debidos a incosistencias.

In [11]:
print(f'Número total de NA en el dataframe: {(col_na:=df_noise.isna().sum()).sum()}')
df_noise_na = pd.DataFrame({'Valores NA': col_na})
df_noise_na

Número total de NA en el dataframe: 0


Valores NA
network_code                   0
receiver_code                  0
receiver_type                  0
receiver_latitude              0
receiver_longitude             0
receiver_elevation_m           0
trace_start_time               0
trace_category                 0

Realizamos comprobaciones similares para los datos de terremotos.

In [12]:
df_earthquakes = df.loc[df.loc[:, 'trace_category'] == 'earthquake_local']
size = len(df_earthquakes)
print(f'len(df_earthquakes)={size}, ratio={size/len(df)*100:.2f}%')
df_earthquakes.head()

len(df_earthquakes)=1030231, ratio=81.40%


network_code receiver_code receiver_type  \
trace_name                                                           
109C.TA_20060723155859_EV           TA          109C            BH   
109C.TA_20061103155652_EV           TA          109C            BH   
109C.TA_20061103161223_EV           TA          109C            BH   
109C.TA_20061114133221_EV           TA          109C            BH   
109C.TA_20061127104640_EV           TA          109C            BH   

                           receiver_latitude  receiver_longitude  \
trace_name                                                         
109C.TA_20060723155859_EV            32.8889           -117.1051   
109C.TA_20061103155652_EV            32.8889           -117.1051   
109C.TA_20061103161223_EV            32.8889           -117.1051   
109C.TA_20061114133221_EV            32.8889           -117.1051   
109C.TA_20061127104640_EV            32.8889           -117.1051   

                           receiver_elevation_m  p_arrival_sample p_status  \
trace_name                                                                   
109C.TA_20060723155859_EV                 150.0             700.0   manual   
109C.TA_20061103155652_EV                 150.0             600.0   manual   
109C.TA_20061103161223_EV                 150.0             500.0   manual   
109C.TA_20061114133221_EV                 150.0             900.0   manual   
109C.TA_20061127104640_EV                 150.0             700.0   manual   

                           p_weight  p_travel_sec  ...  source_magnitude_type  \
trace_name                                         ...                          
109C.TA_20060723155859_EV       0.5     17.080000  ...                     ml   
109C.TA_20061103155652_EV       0.5     16.879999  ...                     mb   
109C.TA_20061103161223_EV       0.5     17.260000  ...                     ml   
109C.TA_20061114133221_EV       0.5     17.280001  ...                     ml   
109C.TA_20061127104640_EV       0.5     18.139999  ...                     ml   

                          source_magnitude_author  \
trace_name                                          
109C.TA_20060723155859_EV                     NaN   
109C.TA_20061103155652_EV                     NaN   
109C.TA_20061103161223_EV                     NaN   
109C.TA_20061114133221_EV                     NaN   
109C.TA_20061127104640_EV                     NaN   

                           source_mechanism_strike_dip_rake  \
trace_name                                                    
109C.TA_20060723155859_EV                               NaN   
109C.TA_20061103155652_EV                               NaN   
109C.TA_20061103161223_EV                               NaN   
109C.TA_20061114133221_EV                               NaN   
109C.TA_20061127104640_EV                               NaN   

                          source_distance_deg source_distance_km  \
trace_name                                                         
109C.TA_20060723155859_EV                0.92             102.09   
109C.TA_20061103155652_EV                0.91             101.34   
109C.TA_20061103161223_EV                0.92             101.87   
109C.TA_20061114133221_EV                0.93             103.26   
109C.TA_20061127104640_EV                0.92             102.48   

                           back_azimuth_deg  \
trace_name                                    
109C.TA_20060723155859_EV             159.3   
109C.TA_20061103155652_EV             281.7   
109C.TA_20061103161223_EV             280.5   
109C.TA_20061114133221_EV             281.6   
109C.TA_20061127104640_EV               4.7   

                                                          snr_db  \
trace_name                                                         
109C.TA_20060723155859_EV  [56.79999924 55.40000153 47.40000153]   
109C.TA_20061103155652_EV  [65.         65.5        61.40000153]   
109C.TA_20061103161223_EV  [37.20000076 42.       

In [13]:
print(f'Número total de NA en el dataframe: {(col_na:=df_earthquakes.isna().sum()).sum()}')
df_earthquakes_na = pd.DataFrame({'Valores NA': col_na})
df_earthquakes_na.loc[df_earthquakes_na.loc[:, 'Valores NA'] != 0, :] # Mostramos solo los que no tienen 0

Número total de NA en el dataframe: 5165058


Valores NA
network_code                              44
p_weight                                 174
s_weight                                 155
source_origin_uncertainty_sec         889937
source_error_sec                      570728
source_gap_deg                        649414
source_horizontal_uncertainty_km      589493
source_depth_km                           49
source_depth_uncertainty_km           660808
source_magnitude_author               780053
source_mechanism_strike_dip_rake     1024203

Podemos comprobar mediante source_id como se distribuyen el número de muestras que hacen referencia al mismo terremoto

In [14]:
earthquake_samples = df_earthquakes.groupby('source_id')
sizes = earthquake_samples.size()
sizes.describe()

count    441705.000000
mean          2.332396
std           2.380903
min           1.000000
25%           1.000000
50%           1.000000
75%           3.000000
max          90.000000
dtype: float64

# Modelo SVM(SVC)

A continuación vamos a desarrollar un modelo para clasificar las señales sísmicas de forma binaria en “earthquake_local” (terremoto local) y “noise” (ruido). Para ello emplearemos SVC (Support Vector Classification), qué es un método muy eficaz para problemas de clasificación binaria.
A pesar de que hay una gran cantidad de algoritmos de clasificación binaria, los SVM(Support Vector Machine) presentan una gran robustez frente a la alta dimensionalidad, es decir, en problemas donde hay un gran número de características (que es nuestro caso pues tenemos 35 atributos). También tiene varias características que pueden llegar a resultar de utilidad como la capacidad de explorar en kernels no lineales, si las relaciones entre características y etiquetas lo fueran.
No obstante, hay situaciones en las que no funcione tan bien; por  ejemplo, si se produce un desbalance significativo entre las clases a predecir (tendrá problemas al clasificar la clase minoritaria) o problemas de escalabilidad con conjunto de datos muy grandes si no se ha realizado un preprocesamiento adecuado.

Comenzamos importando las librerías necesarias:


In [9]:
import pandas as pd
import numpy as np
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import h5py
import os

Accuracy: 0.82


A continuación especificamos los tipos de datos de todos los atributos, para evitar cualquier error relacionado al cargar el conjunto de datos

In [ ]:
# Especificar los tipos de datos para evitar advertencias
dtypes = {
    'network_code': 'str',
    'receiver_code': 'str',
    'receiver_type': 'str',
    'receiver_latitude': 'float64',
    'receiver_longitude': 'float64',
    'receiver_elevation_m': 'float64',
    'p_arrival_sample': 'float64',
    'p_status': 'str',
    'p_weight': 'float64',
    'p_travel_sec': 'float64',
    's_arrival_sample': 'float64',
    's_status': 'str',
    's_weight': 'float64',
    'source_id': 'str',
    'source_origin_time': 'str',
    'source_origin_uncertainty_sec': 'float64',
    'source_latitude': 'float64',
    'source_longitude': 'float64',
    'source_error_sec': 'float64',
    'source_gap_deg': 'float64',
    'source_horizontal_uncertainty_km': 'float64',
    'source_depth_km': 'float64',
    'source_depth_uncertainty_km': 'float64',
    'source_magnitude': 'float64',
    'source_magnitude_type': 'str',
    'source_magnitude_author': 'str',
    'source_mechanism_strike_dip_rake': 'str',
    'source_distance_deg': 'float64',
    'source_distance_km': 'float64',
    'back_azimuth_deg': 'float64',
    'snr_db': 'str',
    'coda_end_sample': 'str',
    'trace_start_time': 'str',
    'trace_category': 'str',
    'trace_name': 'str'
}

Cargamos los metadatos de nuestra base de datos desde el archivo CSV. Para ello se define la ruta del directorio de entrada y el nombre del archivo CSV, luego se carga en un “DataFrame” de pandas.

In [ ]:
# Cargar los datos con tipos de datos especificados
input_dir = '/kaggle/input/stanford-earthquake-dataset-stead/'
metadata_path = 'merge.csv'
df = pd.read_csv(os.path.join(input_dir, metadata_path), dtype=dtypes, low_memory=False)

Ahora seleccionamos las características (que serían los predictores) y la etiqueta (que sería la variable objetivo) que se utilizarán para el modelo:

In [ ]:
# Seleccionar características y etiquetas
features = df[['receiver_latitude', 'receiver_longitude', 'receiver_elevation_m']]
labels = df['trace_category'].apply(lambda x: 1 if x == 'earthquake_local' else 0)


Como hemos mencionado anteriormente el preprocesamiento es un paso crucial, por lo que vamos a escalar las características para normalizar los datos.
En este caso, El "StandardScaler" estandariza las características eliminando la media y escalando a la varianza unitaria.

In [ ]:
# Escalar las características
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)

A  continuación hacemos una división estándar en conjuntos de entrenamiento y prueba, con:


*X_train y y_train:* Datos de entrenamiento.

*X_test y y_test:* Datos de prueba.

*test_size=0.2* indica que el 20% de los datos se utilizarán para pruebas.

*random_state=42*

In [ ]:
# Dividir en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(features_scaled, labels, test_size=0.2, random_state=42)

Para el entrenamiento del modelo, definimos una función empleando *SGDClassifier* en lotes:

*SGDClassifier*: Utiliza la pérdida 'hinge' que es equivalente a un SVM lineal.

*partial_fit*: Permite entrenar el modelo en lotes, adecuado para grandes volúmenes de datos.

*batch_size*: Define el tamaño del lote para cada iteración de entrenamiento.

In [ ]:
# Función para entrenar SVM en lotes utilizando SGDClassifier
def train_svm_in_batches(X_train, y_train, batch_size=10000):
    svm = SGDClassifier(loss='hinge', max_iter=1000, tol=1e-3)
    num_samples = X_train.shape[0]
    for i in range(0, num_samples, batch_size):
        X_batch = X_train[i:i+batch_size]
        y_batch = y_train[i:i+batch_size]
        svm.partial_fit(X_batch, y_batch, classes=np.unique(y_train))
    return svm

Entrenamos el modelo con la función definida anteriormente:

In [ ]:
# Entrenar SVM en lotes
svm_model = train_svm_in_batches(X_train, y_train, batch_size=10000)

Ahora definimos otra función que se encargará de predecir los resultados en el conjunto de prueba en lotes:

*predict*: Realiza predicciones para cada lote de datos.

*y_pred*: Lista que almacena las predicciones de todos los lotes.

In [ ]:
# Predecir en el conjunto de prueba por lotes
def predict_in_batches(model, X_test, batch_size=10000):
    num_samples = X_test.shape[0]
    y_pred = []
    for i in range(0, num_samples, batch_size):
        X_batch = X_test[i:i+batch_size]
        y_pred_batch = model.predict(X_batch)
        y_pred.extend(y_pred_batch)
    return np.array(y_pred)



Finalmente, se predicen los resultados del conjunto de prueba y se evalúa la precisión del modelo

In [ ]:
y_pred = predict_in_batches(svm_model, X_test, batch_size=10000)
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

Como se puede apreciar, la precisión del modelo es 0.82; lo que quiere decir que predecirá correctacmente si se trata de un terremoto o ruido sísmico el 82% de las ocasiones.

Estamos bastante satisfechos, pues consideramos que una precisión superior al 80% aporta bastante credibilidad al modelo. Además, aunque se trata de una clasificación binaria en la que se suele esperar un porcentaje de aciertos elevados, dada la complejidad de los datos  

A continuación, vamos a repetir el modelo empleando los datos ".hdf5" que contienen formas de onda sísmica. Puesto que se trata de datos mucho más pesados y complejos, el modelo tendrá ciertas diferencias notables en el tratamiento de los datos.    

**(WARNING: Dado el gran peso de los datos, la ejecuación de esta celda ronda los 15-20 minutos. Si no se desea comprobar el resultado, ignorar)**

In [7]:
import h5py

def load_waveform_sample(hdf5_path, num_samples=1000):
    with h5py.File(hdf5_path, 'r') as f:
        keys = list(f['data'].keys())
        sampled_keys = np.random.choice(keys, num_samples, replace=False)
        waveforms = [f['data'][key][:] for key in sampled_keys]
    return np.array(waveforms)

# Cargar una muestra de datos del archivo HDF5
waveform_samples = load_waveform_sample(os.path.join(input_dir, waveforms_path), num_samples=1000)

# Asegurarse de que las dimensiones de las formas de onda sean adecuadas para el SVM
# Ejemplo: aplanar las formas de onda si es necesario
waveform_samples_flat = waveform_samples.reshape(waveform_samples.shape[0], -1)

# Seleccionar una muestra correspondiente de características y etiquetas
sample_indices = np.random.choice(len(features_scaled), len(waveform_samples), replace=False)
X_sample = features_scaled[sample_indices]
y_sample = labels.iloc[sample_indices]

# Dividir la muestra en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_sample, y_sample, test_size=0.2, random_state=42)

# Entrenar y evaluar el modelo como se mostró anteriormente
svm_model = train_svm_in_batches(X_train, y_train, batch_size=100)
y_pred = predict_in_batches(svm_model, X_test, batch_size=100)
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy with waveform samples: {accuracy:.2f}')


Accuracy with waveform samples: 0.82
